dandiapi/storage.py

```diff
@@ -0,0 +1,116 @@
+from __future__ import annotations
+
+import json
+from typing import TYPE_CHECKING, Any
+
+from botocore.exceptions import ClientError
+from storages.backends.s3 import S3Storage
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+
+    from mypy_boto3_s3.client import S3Client
+
+
+class DandiS3Storage(S3Storage):
+    """
+    An enhanced S3Storage.
+
+    This class additionally:
+    * Allows unsigned URLs to be generated
+    * Provides an API to tag objects
+    """
+
+    # S3Storage provides this, but doesn't properly annotate it
+    bucket_name: str
+
+    @property
+    def s3_client(self) -> S3Client:
+        return self.connection.meta.client
+
+    def _url_unsigned(self, name: str) -> str:
+        if self.endpoint_url:
+            # TODO: correct URL when inside Docker
+            # Assume only path-style requests are supported, as this is probably MinIO
+            return f'{self.endpoint_url}/{self.bucket_name}/{name}'
+        # https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#virtual-hosted-style-access
+        return f'https://{self.bucket_name}.s3.{self.region_name}.amazonaws.com/{name}'
+
+    def url(
+        self,
+        name: str,
+        *,
+        parameters: Mapping[str, Any] | None = None,
+        expire: int | None = None,
+        http_method: str | None = None,
+        signed: bool = True,
```

Comment:
We should make more use of `signed`, as anything that isn't embargoed can be unsigned.